src/commons/eventsourcing/cmd.rs

```diff
@@ -20,6 +19,95 @@ use crate::commons::{
 /// CommandSummary for use in history.
 pub trait WithStorableDetails: Storable + Send + Sync {
     fn summary(&self) -> CommandSummary;
+
+    /// Create an instance representing the initialisation command/event.
+    ///
+    /// Should not contain any data, so that we can keep this signature
+    /// simple. All relevant data will be in the init event for the command.
+    /// Having this allows us to make a generic pre-0.14.0 AggregateStore
+    /// upgrade implementation.
+    fn make_init() -> Self;
+}
+
+//------------ InitCommand ---------------------------------------------------
+
+/// The InitCommand is used to create an aggregate.
+///
+/// It should be storable in the same way as normal commands, sent to this
+/// aggregate type so that they can use the same kind of ProcessedCommand
+/// and CommandHistoryRecord
+pub trait InitCommand: fmt::Display + Send + Sync {
+    /// Identify the type of storable component for this command. Commands
+    /// may contain short-lived things (e.g. an Arc<Signer>) or even secrets
+    /// which should not be persisted.
+    type StorableDetails: WithStorableDetails;
+
+    /// Identifies the aggregate, useful when storing and retrieving the event.
+    fn handle(&self) -> &MyHandle;
+
+    /// The actor who sent the command. There is no default so as to avoid
+    /// accidentally attributing a command by a user instead as if it were an
+    /// internal command by Krill itself.
+    fn actor(&self) -> &str;
+
+    /// Get the storable information for this command
+    fn store(&self) -> Self::StorableDetails;
+}
+
+//------------ SentInitCommand -----------------------------------------------
+
+/// Convenience wrapper so that implementations can just implement
+/// ['InitCommandDetails'] and leave the id and version boilerplate.
+pub struct SentInitCommand<I: InitCommandDetails> {
+    handle: MyHandle,
+    details: I,
+    actor: String,
+}
+
+impl<I: InitCommandDetails> InitCommand for SentInitCommand<I> {
+    type StorableDetails = I::StorableDetails;
+
+    fn handle(&self) -> &MyHandle {
+        &self.handle
+    }
+
+    fn actor(&self) -> &str {
+        &self.actor
+    }
+
+    fn store(&self) -> Self::StorableDetails {
+        self.details.store()
+    }
+}
+
+impl<I: InitCommandDetails> SentInitCommand<I> {
+    pub fn new(id: &MyHandle, details: I, actor: &Actor) -> Self {
+        SentInitCommand {
+            handle: id.clone(),
+            details,
+            actor: actor.to_string(),
+        }
+    }
+
+    pub fn into_details(self) -> I {
+        self.details
+    }
+}
+
+impl<I: InitCommandDetails> fmt::Display for SentInitCommand<I> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "initialise '{}'", self.handle)
```

Comment:
Other implementations like this call `self.store.fmt()`. Is it right that that is different here?

Reply:
Yeah, this is correct. This is one level higher - those other implementations are on the level of the "command details".